libdevcrypto/gm/sm2/sm2.cpp

```diff
@@ -21,353 +21,199 @@
-    if (!pri)
-    {
-        CRYPTO_LOG(ERROR) << "[SM2::genKey] Error Of Output SM2 Private key";
-        goto err;
+    point = EC_POINT_new(sm2Group);
+    if (!point) {
+      CRYPTO_LOG(ERROR) << "[SM2::sign] ERROR of EC_POINT_new";
+      goto done;
     }
-    privateKey = pri;
-    // LOG(DEBUG)<<"SM2 PrivateKey:"<<privateKey;
 
-    pub = EC_POINT_point2hex(
-        sm2Group, EC_KEY_get0_public_key(sm2Key), POINT_CONVERSION_UNCOMPRESSED, NULL);
-    if (!pub)
-    {
-        CRYPTO_LOG(ERROR) << "[SM2::genKey] Error Of Output SM2 Public key";
-        goto err;
+    if (!EC_POINT_mul(sm2Group, point, privateKey, NULL, NULL, NULL)) {
+      CRYPTO_LOG(ERROR) << "[SM2::sign] ERROR of EC_POINT_mul";
+      goto done;
     }
-    publicKey = pub;
-    lresult = true;
-    // LOG(DEBUG)<<"SM2 PublicKey:"<<publicKey;
-err:
-    if (pri)
-        OPENSSL_free(pri);
-    if (pub)
-        OPENSSL_free(pub);
-    if (sm2Group)
-        EC_GROUP_free(sm2Group);
-    if (sm2Key)
-        EC_KEY_free(sm2Key);
-    return lresult;
-}
-
-string SM2::getPublicKey()
-{
-    publicKey = publicKey.substr(2, 128);
-    publicKey = strlower((char*)publicKey.c_str());
-    return publicKey;
-}
-
-string SM2::getPrivateKey()
-{
-    privateKey = strlower((char*)privateKey.c_str());
-    return privateKey;
-}
-
-bool SM2::sign(
-    const char* originalData, int originalDataLen, const string& privateKey, string& r, string& s)
-{
-    // originalDataLen:"<<originalDataLen<<" Sign privateKey:"<<privateKey;
-
-    bool lresult = false;
-    SM3_CTX sm3Ctx;
-    EC_KEY* sm2Key = NULL;
-    unsigned char zValue[SM3_DIGEST_LENGTH];
-
-    size_t zValueLen;
-    ECDSA_SIG* signData = NULL;
-    char* rData = NULL;
-    char* sData = NULL;
-    string str = "";   // big int data
-    string _str = "";  // big int swap data
-    int _size = 0;     // big int padding size
-    BIGNUM start;
-    BIGNUM* res = NULL;
-    BN_CTX* ctx = NULL;
-    BN_init(&start);
-    ctx = BN_CTX_new();
-
-    res = &start;
-    BN_hex2bn(&res, (const char*)privateKey.c_str());
-    sm2Key = EC_KEY_new_by_curve_name(NID_sm2);
-    EC_KEY_set_private_key(sm2Key, res);
-
-    zValueLen = sizeof(zValue);
-    if (!sm2GetZ(privateKey, (const EC_KEY*)sm2Key, zValue, zValueLen))
-    {
-        CRYPTO_LOG(ERROR) << "[SM2::sign] Error Of Compute Z";
-        goto err;
+    if (!EC_KEY_set_public_key(sm2Key, point)) {
+      CRYPTO_LOG(ERROR) << "[SM2::sign] ERROR of EC_KEY_set_public_key";
+      goto done;
     }
-    // SM3 Degist
-    SM3_Init(&sm3Ctx);
-    SM3_Update(&sm3Ctx, zValue, zValueLen);
-    SM3_Update(&sm3Ctx, originalData, originalDataLen);
-    SM3_Final(zValue, &sm3Ctx);
-
-    signData = ECDSA_do_sign_ex(zValue, zValueLen, NULL, NULL, sm2Key);
-    if (signData == NULL)
-    {
-        CRYPTO_LOG(ERROR) << "[SM2::sign] Error Of SM2 Signature";
-        goto err;
+    sig =
+        sm2_do_sign(sm2Key, EVP_sm3(), (const uint8_t *)userId, strlen(userId),
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Does not handle strings that are not \0-terminated; if given one it may perform an over-read (it could cause a crash if unprotected) (CWE-126).  ](https://app.codacy.com/gh/FISCO-BCOS/FISCO-BCOS/pullRequest?prid=5347884)